Automatically judge users in massjoins if they're bots and use wizard to ban suspected bots directly

The-personified-devil opened the pull request.

Commit:
Add automatic bot judging and direct banning

Signed-off-by: Leonhard Saam <leonhard60-git@yahoo.com>

Yuhanun commented on src/cogs/rules.py:
```diff
@@ -20,8 +20,7 @@ def __init__(self, bot):
 
         self.massjoin_detect = True
-
-        self.next_massjoin_notif = time.time()
+        self.massjoin_active = False
```
what's the difference?

The-personified-devil replied:
The old code used to wait a certain amount after it detected a massjoin and only call out a massjoin again after the timeout was finished.
The new code blocks by waiting for reactions and after that everything gets reset.

The new flag also determines behavior specific to after a massjoin has been the detected, like not removing old entries in the recent_joins list etc.

Yuhanun commented on src/cogs/rules.py:
```diff
@@ -59,19 +60,79 @@ async def on_member_join(self, member):
+                if current_time - x["join_time"] <= self.file["massjoin_window"]
+            ]
+
+        self._recent_joins.append({"join_time": current_time, "id": member.id,
+            "assumed_bot": (
+                ( member.default_avatar_url == member.avatar_url if self.file["massjoin_default_pfp"] else True )
+                and ( time.time() - member.created_at.timestamp() < self.file["massjoin_min_acc_age_val"]
+                    if self.file["massjoin_min_acc_age"] else True ) )
+            })
```
format this, every key on its own line + goddamn that condition make it into a function.

"assumed_bot": is_assumed_bot(member)

The-personified-devil replied:
Should the first key be on it's own line too?

Yuhanun replied:
yes, the `{` should be next to the `(`, but join_time sh ould be on next line, indented from `self`

Yuhanun replied:
```py
self._recent_joins.append({
    "join_time": ...,
    "assumed_bot": is_assumed_bot(member),
})```

Yuhanun commented on src/cogs/rules.py:
```diff
@@ -59,19 +60,79 @@ async def on_member_join(self, member):
+        if join_amount >= self.file["massjoin_amount"] and not self.massjoin_active:
+            self.massjoin_active = True
+
+            cross = "❎"
+            checkmark = "✅"
```
make these configurable

Yuhanun commented on src/cogs/rules.py:
```diff
@@ -59,19 +60,79 @@ async def on_member_join(self, member):
+            await msg.add_reaction(cross)
+            await msg.add_reaction(checkmark)
+
+            def _check(reaction, user, msg):
```
shadowing variable `msg`, rename it to `reaction_msg`

Yuhanun commented on src/cogs/rules.py:
```diff
@@ -59,19 +60,79 @@ async def on_member_join(self, member):
+                await msg.reply("Not taking any action and resetting the join detection")
+
+            if reaction.emoji == checkmark:
+                reply = await msg.reply("Users assumed to be bots:\n" + ",\n".join(map(lambda x: f"<@{x['id']}>",
+                    filter(lambda x: x["assumed_bot"], self._recent_joins)))
+                    + "\nUsers assumed to not be bots:\n" + ",\n".join(map(lambda x: f"<@{x['id']}>",
+                    filter(lambda x: not x["assumed_bot"], self._recent_joins))))
```
This might not fit in a message, use an IO file and send the file

The-personified-devil replied:
This has the disadvantage of not having the users in the form of pings allowing to access them and get the id.
Adding to this, 2000 characters can actually be quite a few pings.
Since that doesn't ensure everything fits into the message, I'm going to add something to auto-split it into multiple messages.

Yuhanun replied:
that's fine then, multiple messages is ok

Yuhanun commented on src/cogs/rules.py:
```diff
@@ -59,19 +60,79 @@ async def on_member_join(self, member):
+
+                if reaction.emoji == checkmark:
+                    for user in self._recent_joins:
+                        if user["assumed_bot"]:
```
if not use[...]: continue
await member.guild.ban...

Yuhanun replied:
do put continue on a newline tho

The-personified-devil replied:
I don't fully understand what I'm intended to do here.

Yuhanun replied:
flip the condition
```
if x:
   continue
do_something()
```

instead of

```
if not x:
    do_something()
    ```

Yuhanun commented on src/cogs/rules.py:
```diff
@@ -59,19 +60,79 @@ async def on_member_join(self, member):
+                    await reply.reply("Not banning any users and resetting the join detection")
+                    await reply.clear_reactions()
+
+                if reaction.emoji == checkmark:
```
if reaction.emoji == cross: return await ctx.send("Okay, I'll leave it up to you!")

The-personified-devil replied:
I prefer my version of the text as it also mentions the fact that the join detection is reset, so fully armed again.
If your goal is actually to not reset the join detection, then I'd also have to provide away to do so.
I actually forgot to delete the `clear_reactions()` there, since it automatically done at the end of the scope.
But I can remove the reply if you want.

Yuhanun replied:
nah, then it's okay as-is if it works

Commit:
Add automatic splitting for long massjoin messages

Signed-off-by: Leonhard Saam <leonhard.saam@yahoo.com>

Commit:
Fix multiple issues

- Make the yes/no reaction configurable
- Remove unneeded `clear_reactions()`
- Fix the formatting of the member entry into the list
- Fixed shadowing `msg` in `_check()`

Signed-off-by: Leonhard Saam <leonhard.saam@yahoo.com>

Yuhanun reviewed (approved):
LGTM
